pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -81,126 +90,94 @@ func NewDRAPluginClient(pluginName string) (drapb.NodeClient, error) {
 		return nil, fmt.Errorf("plugin name %s not found in the list of registered DRA plugins", pluginName)
 	}
 
+	conn, err := existingPlugin.getOrCreateGRPCConn()
+	if err != nil {
+		return nil, err
+	}
+
 	return &draPluginClient{
-		pluginName:        pluginName,
-		addr:              draAddr(existingPlugin.endpoint),
-		nodeClientCreator: newNodeClient,
+		pluginName:      pluginName,
+		plugin:          existingPlugin,
+		versionedClient: detectVersion(conn),
 	}, nil
 }
 
 func (r *draPluginClient) NodePrepareResources(
 	ctx context.Context,
-	req *drapb.NodePrepareResourcesRequest,
+	req *drapbv1alpha3.NodePrepareResourcesRequest,
 	opts ...grpc.CallOption,
-) (resp *drapb.NodePrepareResourcesResponse, err error) {
+) (response *drapbv1alpha3.NodePrepareResourcesResponse, err error) {
 	logger := klog.FromContext(ctx)
 	logger.V(4).Info(log("calling NodePrepareResources rpc"), "request", req)
-	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", resp, "err", err)
+	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", response, "err", err)
 
-	if r.nodeClientCreator == nil {
-		return nil, errors.New("failed to call NodePrepareResources. nodeClientCreator is nil")
-	}
-
-	nodeClient, nodeClientOld, closer, err := r.nodeClientCreator(r.addr)
+	conn, err := r.plugin.getOrCreateGRPCConn()
 	if err != nil {
 		return nil, err
 	}
-	defer closer.Close()
 
 	ctx, cancel := context.WithTimeout(ctx, PluginClientTimeout)
 	defer cancel()
 
-	resp, err = nodeClient.NodePrepareResources(ctx, req)
-	if err != nil {
-		status, _ := grpcstatus.FromError(err)
-		if status.Code() == grpccodes.Unimplemented {
-			// Fall back to the older gRPC API.
-			resp = &drapb.NodePrepareResourcesResponse{
-				Claims: make(map[string]*drapb.NodePrepareResourceResponse),
-			}
-			err = nil
-			for _, claim := range req.Claims {
-				respOld, errOld := nodeClientOld.NodePrepareResource(ctx,
-					&drapbv1alpha2.NodePrepareResourceRequest{
-						Namespace:      claim.Namespace,
-						ClaimUid:       claim.Uid,
-						ClaimName:      claim.Name,
-						ResourceHandle: claim.ResourceHandle,
-					})
-				result := &drapb.NodePrepareResourceResponse{}
-				if errOld != nil {
-					result.Error = errOld.Error()
-				} else {
-					result.CDIDevices = respOld.CdiDevices
-				}
-				resp.Claims[claim.Uid] = result
-			}
+	if r.versionedClient.version == "v1alpha3" {
+		response, err = r.versionedClient.v1alpha3NodePrepareResources(ctx, conn, req)
+		return
+	}
+
```

Comment:
Can we have an explicit `if r.versionedClient.version == "v1alpha2"`here? I'd like to explicitly return an error if it's not one of these two (to future proof things.

I'd also prefer if the conversion logic below and above was pulled into a function rather than inlined, so that this will read:
```
if r.versionedClient.version == "v1alpha3" {
	response, err = r.versionedClient.v1alpha3NodePrepareResources(ctx, conn, req)
	if err != nil {
		return nil, fmt.Errorf("error calling v1alpha3.NodePrepareResources: %v", err)
	}
	return response, nil
}
if r.versionedClient.version == "v1alpha3" {
	response, err = r.versionedClient.v1alpha2NodePrepareResource(ctx, conn, req)
	if err != nil {
		return nil, fmt.Errorf("error calling v1alpha2.NodePrepareResource: %v", err)
	}
	convertedResponse, err = r.versionedClient.convertV1alpha2NodePrepareResourceResponseToV1alpha3NodePrepareResourcesResponse(response)
	if err != nil {
		return nil, fmt.Errorf("error converting v1alpha2.NodePrepareResourceResponse to v1alpha3.NodePrepareResourcesResponse: %v", err)
	}
	return convertedResponse, nil
}

return nil, fmt.Errorf("unknown plugin API version")
```

Reply:
Given my other comment [here](https://github.com/kubernetes/kubernetes/pull/118619#discussion_r1265626550), this would actually turn into:
```
if r.pluginAPIVersion == "v1alpha3" {
	nodeClient := drapbv1alpha3.NewNodeClient(conn)
	response, err = nodeClient.NodePrepareResources(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("error calling v1alpha3.NodePrepareResources: %v", err)
	}
	return response, nil
}
if r.pluginAPIVersion == "v1alpha2" {
	nodeClient := drapbv1alpha2.NewNodeClient(conn)
	response, err = nodeClient.NodePrepareResource(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("error calling v1alpha2.NodePrepareResource: %v", err)
	}
	convertedResponse, err = r.versionedClient.convertV1alpha2NodePrepareResourceResponseToV1alpha3NodePrepareResourcesResponse(response)
	if err != nil {
		return nil, fmt.Errorf("error converting v1alpha2.NodePrepareResourceResponse to v1alpha3.NodePrepareResourcesResponse: %v", err)
	}
	return convertedResponse, nil
}

return nil, fmt.Errorf("unknown plugin API version")
```

Reply:
done in 832264ac0d4